### ISSUE TITLE
the parser is singleton or instance? 

### ISSUE THREAD

**Issue by xp44mm** [AngleSharp/AngleSharp#885]
We will use `parse` to parse a lots of text that is html string:

```F#
let doc1 = parse inp1;;
let doc2 = parse inp2;;
let doc3 = parse inp3;;
...
```

The following two `parse` implement codes, which is right?

parser is singleton:

```F#
// We can think of it like a tab in a standard browser.
let context = BrowsingContext.New(Configuration.Default)    
let htmlParser = context.GetService<IHtmlParser>()
let parse (source:string) =
    htmlParser.ParseDocument(source)
```

and parser is instance:

```F#
let parse (source:string) =
    // We can think of it like a tab in a standard browser.
    let context = BrowsingContext.New(Configuration.Default)    
    let htmlParser = context.GetService<IHtmlParser>()
    htmlParser.ParseDocument(source)
```

**Comment by FlorianRappl** [comment]
You wouldn't use a single tab in your browser to read pages in parallel (i.e., you can't), so the same applies to a `BrowsingContext`.

For subsequent page requests re-using the `BrowserContext` is fine. In general, you could achieve an efficient re-use in your application by pooling `BrowsingContext` instances.

I hope that helps!

**Comment by Tyler-V** [comment]
@FlorianRappl I've been looking around are there any suitable examples on pooling BrowsingContexts? What I am running into is creating them at runtime using the thread pool as I have dynamic proxies and random user agent's I need to set in the HttpClientHandler and DefaultHttpRequester.

I find that I often run into null reference exceptions.

![image](https://user-images.githubusercontent.com/3893234/102704428-16f2f880-4249-11eb-96ce-5e70d22437c9.png)

I can open a new issue if you like
